compiler/parser/parser.cpp

```diff
@@ -413,6 +416,100 @@ static UniqueString cleanLocalPath(UniqueString path) {
   return chpl::UniqueString::get(gContext, str);
 }
 
+static void gatherStdModuleNamesInDir(std::string dir,
+                                      std::set<UniqueString>& modNames) {
+  std::error_code EC;
+  llvm::sys::fs::directory_iterator I(dir, EC);
+  llvm::sys::fs::directory_iterator E;
+
+  std::set<std::string> moduleNamesHere;
+  std::set<std::string> subDirsHere;
+  while (true) {
+    if (I == E || EC) {
+      break;
+    }
+    // consider the file
+    llvm::StringRef fileName = llvm::sys::path::filename(I->path());
+    bool isdir = I->type() == llvm::sys::fs::file_type::directory_file;
+    if (isdir) {
+      subDirsHere.insert(fileName.str());
+    } else {
+      llvm::StringRef fileExt = llvm::sys::path::extension(fileName);
+      llvm::StringRef fileStem = llvm::sys::path::stem(fileName);
+      if (fileExt.equals(".chpl")) {
+        moduleNamesHere.insert(fileStem.str());
+      }
+    }
+
+    I.increment(EC);
+  }
+
+  if (EC) {
+    USR_FATAL("error in directory traversal of %s", dir.c_str());
+  }
+
+  // Consider the gathered module names & add them to the returned set.
+  for (const auto& name : moduleNamesHere) {
+    modNames.insert(UniqueString::get(gContext, name));
+  }
+
+  // Consider the subdirs. Visit any subdir that does not have
+  // the same name as a module here. This is meant to exclude submodules
+  // stored in different files. For example, if we have
+  //   SortedSet/
+  //   SortedSet.chpl
+  // then we should not traverse into SortedSet/ under the assumption
+  // that it is storing only submodules.
+  for (const auto& subdir : subDirsHere) {
+    if (moduleNamesHere.count(subdir) == 0) {
+      std::string subPath = dir + "/" + subdir;
+      gatherStdModuleNamesInDir(subPath, modNames);
+    } else {
+    }
+  }
+}
+
+static std::set<UniqueString> gatherStdModuleNames() {
+  // compute $CHPL_HOME/modules
+  const auto& bundledPath = chpl::parsing::bundledModulePath(gContext);
+  std::string modulesDir = bundledPath.str();
+
+  std::set<UniqueString> modNames;
+  gatherStdModuleNamesInDir(modulesDir + "/dists", modNames);
+  gatherStdModuleNamesInDir(modulesDir + "/internal", modNames);
+  gatherStdModuleNamesInDir(modulesDir + "/layouts", modNames);
+  // skip minimal
+  gatherStdModuleNamesInDir(modulesDir + "/packages", modNames);
+  gatherStdModuleNamesInDir(modulesDir + "/standard", modNames);
+
+  // leave out Treap since it is an 'include module' for SortedSet
+  modNames.erase(UniqueString::get(gContext, "Treap"));
+
+  return modNames;
+}
+
+static std::vector<UniqueString> gatherStdModulePaths() {
```

Comment:
It's because our standard/internal modules sometimes have multiple implementations of a module that are determined by the module search path. E.g.

```
modules/internal/comm
modules/internal/comm/ugni
modules/internal/comm/ugni/NetworkAtomicTypes.chpl
modules/internal/comm/ofi
modules/internal/comm/ofi/NetworkAtomicTypes.chpl
```

here, `NetworkAtomicTypes` is a standard module that should be available. But, which one we load depends on the compiler configuration, which impacts the module search path.

Reply:
I'll add a comment to this effect.

Reply:
I also updated the PR message to clarify why it is this way.